src/Console/CheckoutCommand.php

```diff
@@ -11,50 +14,203 @@
+            $oldLogger = $repo->getLogger();
+            $repo->setLogger(new NullLogger());
+            try {
+                $result = $repo->run('rev-parse', ['--quiet', '--verify', 'origin/' . $majorBranch]);
+            } catch (ProcessException $e) {
+                // no-op, this happens if the result is null and the repo is in debug mode.
+            } finally {
+                $repo->setLogger($oldLogger);
+            }
+
+            // If major branch doesn't exist, get the last available minor branch
+            if ($result === null) {
+                // Get all branches and remove the remote name from the front
+                $allBranches = array_map(
+                    fn($branchName) => trim(str_replace('origin/', '', $branchName)),
+                    explode(PHP_EOL, $repo->run('branch', ['-r']))
+                );
+
+                // Exclude branches not in the correct major release
+                $allBranches = array_filter($allBranches, fn($branchName) => preg_match('/^' . preg_quote($majorBranch, '/') . '\.\d+$/', $branchName));
+
+                // Use the numerically largest option
+                sort($allBranches, SORT_NATURAL);
+                $data['versionmap'][$version] = $branch = end($allBranches) ?: null;
+            } else {
+                // No change - but still note the branch so we can update the repo data
+                $branch = $majorBranch;
+            }
+
+            if ($branch === null) {
+                trigger_error("No valid branch available for $name in version $version", E_USER_WARNING);
+            } elseif ($updateRepoData) {
+                $repo->run('checkout', ['--force', $branch]);
+                $repo->run('reset', ['--hard', 'origin/' . $branch]);
+            }
+        }
+    }
+
+    /**
+     * Writes a version mapping array to disk as JSON, so other areas of the code can use it.
+     */
+    private function writeMapToDisk(array $map, array $config, OutputInterface $output)
+    {
+        $output->writeln('Writing version map to disk');
+        $filePath = Config::configPath($config['paths']['versionmap']);
+        $success = file_put_contents(
+            $filePath,
+            json_encode($map, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE)
+        );
+        if ($success === false) {
+            throw new RuntimeException('Failed to write version map to disk. Attempted to write to ' . $filePath);
+        }
+    }
+
+    /**
+     * Builds a map of which Silverstripe CMS major release matches which major release of a module.
+     */
+    private function buildVersionMap(array $config, OutputInterface $output)
+    {
+        $output->writeln('Making new version map');
+        $map = [];
+        $versions = array_keys($config['versions']);
+
+        foreach ($versions as $version) {
+            $modulesUrl = "https://raw.githubusercontent.com/silverstripe/supported-modules/$version/modules.json";
```

Comment:
This requires the supported-modules PRs to be merged to work correctly - but if you want to test this locally in the mean time you can replace that one URL with this code:

```php
switch ($version) {
    case '5':
        $modulesUrl = 'https://raw.githubusercontent.com/silverstripe/supported-modules/e9c9d44a6833bda035d6a3182effd167ab499549/modules.json';
        break;
    case '4':
        $modulesUrl = 'https://raw.githubusercontent.com/silverstripe/supported-modules/daa19e8c279f599088ba12b2a91b9a51cfbc5eb2/modules.json';
        break;
    case '3':
        $modulesUrl = 'https://raw.githubusercontent.com/silverstripe/supported-modules/30ceab854f964f4488f2834b80fc4d56f464e9c2/modules.json';
        break;
}
```